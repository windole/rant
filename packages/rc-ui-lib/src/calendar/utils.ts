/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
