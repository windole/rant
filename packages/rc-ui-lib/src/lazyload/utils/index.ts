import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
