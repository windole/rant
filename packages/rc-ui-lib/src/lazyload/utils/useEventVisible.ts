import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
