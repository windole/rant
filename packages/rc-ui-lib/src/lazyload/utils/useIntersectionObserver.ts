import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
