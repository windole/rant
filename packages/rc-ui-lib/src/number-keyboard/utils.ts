import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
