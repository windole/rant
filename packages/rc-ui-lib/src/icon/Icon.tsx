import React, { useContext, useMemo } from 'react';
import classnames from 'classnames';
import { addUnit } from '../utils';
import { IconProps } from './PropsType';
import Badge from '../badge';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

function isImage(name?: string): boolean {
  return name ? name.indexOf('/') !== -1 : false;
}

const Icon: React.FC<IconProps> = (props) => {
  const { iconPrefix } = useContext(ConfigProviderContext);

  const { dot, tag, name, className, onClick, onTouchStart } = props;

  const imageIcon = isImage(name);

  const classPrefix = useMemo(
    () => props.classPrefix || iconPrefix,
    [props.classPrefix, iconPrefix],
  );

  return (
    <Badge
      dot={dot}
      tag={tag}
      className={classnames(className, classPrefix, imageIcon ? '' : `${classPrefix}-${name}`)}
      style={{
        color: props.color,
        fontSize: addUnit(props.size),
        ...props.style,
      }}
      onClick={onClick}
      onTouchStart={onTouchStart}
      {...props.badge}
    >
      {props?.children}
      {imageIcon && <img className={classnames('rc-icon__image')} src={name} alt={name} />}
    </Badge>
  );
};

Icon.defaultProps = {
  tag: 'i',
};

export default Icon;
