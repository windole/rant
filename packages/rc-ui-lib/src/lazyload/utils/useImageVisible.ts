import { useCallback, useState } from 'react';
import { addCssClassName, loadImage, setImage } from '.';
import { IntersectionObserverEntryType, ObserverOptions } from '../PropsType';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useImageVisible = (
  observerOptions: ObserverOptions = defaultObserverOptions,
  errorImage: string,
): [(node: Element) => void, boolean] => {
  const [isLoaded, setIsLoaded] = useState(false);

  const handleIntersection = useCallback(
    (entries: IntersectionObserverEntryType[]) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLImageElement | HTMLElement;
          const url = target.dataset.src || '';
          loadImage(url)
            .catch(() => {
              setImage(target, errorImage);
              addCssClassName(target, 'rc-lazyload__error');
              target.setAttribute('lazy', 'error');
            })
            .then((imagePath: string) => {
              setImage(target, imagePath);
              setIsLoaded(true);
              addCssClassName(target, 'rc-lazyload__loaded');
              target.setAttribute('lazy', 'loaded');
              target.removeAttribute('data-src');
            });
        });
    },
    [setIsLoaded, isLoaded],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, isLoaded];
};

export default useImageVisible;
