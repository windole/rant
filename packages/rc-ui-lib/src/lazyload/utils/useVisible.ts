import { useCallback, useEffect, useState } from 'react';
import { addCssClassName } from '.';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useVisible = (
  initialVisible: boolean,
  observerOptions = defaultObserverOptions,
): [(node: Element) => void, boolean] => {
  const [visible, setVisible] = useState(initialVisible);

  const forceVisible = () => {
    setVisible(true);
  };

  useEffect(() => {
    if (initialVisible) {
      forceVisible();
    }
  }, [initialVisible]);

  const handleIntersection = useCallback(
    (entries) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLElement;
          addCssClassName(target, 'rc-lazyload__loaded');
          setVisible(true);
        });
    },
    [setVisible, visible],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, visible];
};

export default useVisible;
