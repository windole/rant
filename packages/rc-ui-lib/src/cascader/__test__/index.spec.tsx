import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { sleep } from '../../../tests/utils';
import Icon from '../../icon';
import Cascader from '..';

describe('Cascader', () => {
  let wrapper;
  let spyConsole: jest.SpyInstance;
  beforeEach(() => {
    spyConsole = jest.spyOn(console, 'error');
    spyConsole.mockImplementation(() => {
      return null;
    });
  });

  afterEach(() => {
    wrapper.unmount();
    spyConsole.mockRestore();
    jest.restoreAllMocks();
  });

  const options = [
    {
      text: '浙江省',
      value: '330000',
      children: [{ text: '杭州市', value: '330100' }],
    },
    {
      text: '江苏省',
      value: '320000',
      children: [{ text: '南京市', value: '320100' }],
    },
  ];

  it('should emit change event when active option changed', async () => {
    const onChange = jest.fn();
    wrapper = mount(<Cascader options={options} onChange={onChange} />);

    await sleep();
    wrapper.find('.rc-cascader__option').at(0).simulate('click');

    const firstOption = options[0];
    expect(onChange.mock.calls[0][0]).toEqual({
      value: firstOption.value,
      tabIndex: 0,
      selectedOptions: [firstOption],
    });

    await sleep();
    wrapper.find('.rc-cascader__options').at(1).find('.rc-cascader__option').simulate('click');

    const secondOption = options[0].children[0];
    expect(onChange.mock.calls[1][0]).toEqual({
      value: secondOption.value,
      tabIndex: 1,
      selectedOptions: [firstOption, secondOption],
    });
  });

  it('will not emit onChange event when using disabled prop', async () => {
    const items = [
      {
        text: '广东省',
        value: '310000',
        children: [{ text: '深圳市', value: '310100' }],
        disabled: true,
      },
    ];
    const onChange = jest.fn();
    wrapper = mount(<Cascader options={items} onChange={onChange} />);
    await sleep();
    wrapper.find('.rc-cascader__option').at(0).simulate('click');
    expect(onChange).not.toHaveBeenCalled();
  });

  it('should emit finish event when all options is selected', async () => {
    const option = { value: '1', text: 'foo' };
    const onFinish = jest.fn();
    wrapper = mount(<Cascader options={[option]} onFinish={onFinish} />);
    await sleep();
    wrapper.find('.rc-cascader__option').simulate('click');

    expect(onFinish.mock.calls[0][0]).toEqual({
      value: option.value,
      tabIndex: 0,
      selectedOptions: [option],
    });
  });

  it('should emit close event when close icon is clicked', () => {
    const onClose = jest.fn();
    wrapper = mount(<Cascader onClose={onClose} />);
    wrapper.find(Icon).props().onClick();
    expect(onClose).toHaveBeenCalled();
  });

  it('should not render close icon when closeable is false', () => {
    wrapper = mount(<Cascader closeable={false} />);
    expect(wrapper.find('.rc-cascader__close-icon').exists()).toBeFalsy();
  });

  it('should change close icon when close-icon prop is string', () => {
    wrapper = mount(<Cascader closeIcon="success" />);
    expect(wrapper.find(Icon).props().name).toEqual('success');
  });

  it('should change close icon when close-icon is ValidElement', () => {
    wrapper = mount(<Cascader closeIcon={<Icon name="shop-o" />} />);
    expect(wrapper.find('i.van-icon-shop-o').exists()).toBeTruthy();
  });

  it('should change close icon when close-icon is not ValidElement', () => {
    wrapper = mount(<Cascader closeIcon={123} />);
    expect(wrapper.find('i.van-icon-shop-o').exists()).toBeFalsy();
  });

  test('should render title slot correctly', () => {
    wrapper = mount(<Cascader title="Custom Title" />);
    expect(wrapper.find('.rc-cascader__title').html()).toMatchSnapshot();
  });

  test('should render option correctly when using optionRender prop', async () => {
    const optionItem = { value: '1', text: 'foo' };
    wrapper = mount(
      <Cascader
        options={[optionItem]}
        optionRender={({ option }) => `Custom Option ${option.text}`}
      />,
    );
    await sleep();
    expect(wrapper.find('.rc-cascader__option').text()).toEqual('Custom Option foo');
  });

  test('should render option correctly when using defaultValue prop', async () => {
    Date.now = jest.fn(() => 1482363367071);
    wrapper = mount(<Cascader options={options} defaultValue={['330000']} />);
    await sleep();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should emit onClickTab event correctly', async () => {
    const onClickTab = jest.fn();
    wrapper = mount(<Cascader options={options} onClickTab={onClickTab} />);

    await wrapper.find('.rc-tab').simulate('click');
    expect(onClickTab).toHaveBeenCalled();
  });

  test('when value change', async () => {
    wrapper = mount(<Cascader options={options} defaultValue={['330000']} />);
    await wrapper.setProps({ value: ['350000'] });
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(spyConsole).toHaveBeenLastCalledWith(
      Error(
        'Cascader: unable to match options correctly, Please check value or defaultValue props.',
      ),
    );
  });
});
