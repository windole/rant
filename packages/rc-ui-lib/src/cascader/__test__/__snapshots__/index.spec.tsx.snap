// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cascader should render option correctly when using defaultValue prop 1`] = `
<Cascader
  closeIcon="cross"
  closeable={true}
  defaultValue={
    Array [
      "330000",
    ]
  }
  options={
    Array [
      Object {
        "children": Array [
          Object {
            "text": "杭州市",
            "value": "330100",
          },
        ],
        "text": "浙江省",
        "value": "330000",
      },
      Object {
        "children": Array [
          Object {
            "text": "南京市",
            "value": "320100",
          },
        ],
        "text": "江苏省",
        "value": "320000",
      },
    ]
  }
  swipeable={true}
>
  <div
    className="rc-cascader"
  >
    <div
      className="rc-cascader__header"
    >
      <h2
        className="rc-cascader__title"
      />
      <Icon
        className="rc-cascader__close-icon"
        name="cross"
        onClick={[Function]}
        tag="i"
      >
        <Badge
          className="rc-cascader__close-icon van-icon van-icon-cross"
          onClick={[Function]}
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper rc-cascader__close-icon van-icon van-icon-cross"
            onClick={[Function]}
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
    <ForwardRef
      active={0}
      animated={true}
      className="rc-cascader__tabs"
      duration={300}
      ellipsis={true}
      lazyRender={true}
      offsetTop={0}
      onClickTab={[Function]}
      swipeThreshold={0}
      swipeable={true}
      type="line"
    >
      <div
        className="rc-cascader__tabs rc-tabs rc-tabs--line"
      >
        <div
          className="rc-tabs__wrap rc-tabs__wrap--scrollable"
        >
          <div
            className="rc-tabs__nav rc-tabs__nav--line rc-tabs__nav--complete"
            role="tablist"
            style={
              Object {
                "background": undefined,
                "borderColor": undefined,
              }
            }
          >
            <ForwardRef
              className="rc-cascader__tab"
              isActive={true}
              key=".$0"
              onClick={[Function]}
              scrollable={true}
              showZeroBadge={true}
              title="浙江省"
              type="line"
            >
              <div
                aria-selected={true}
                className="rc-tab rc-tab--active rc-cascader__tab"
                onClick={[Function]}
                style={Object {}}
              >
                <span
                  className="rc-tab__text"
                >
                  浙江省
                </span>
              </div>
            </ForwardRef>
            <div
              className="rc-tabs__line"
              style={
                Object {
                  "backgroundColor": undefined,
                }
              }
            />
          </div>
        </div>
        <TabsContent
          animated={true}
          count={1}
          currentIndex={0}
          duration={300}
          inited={true}
          lazyRender={true}
          onChange={[Function]}
          swipeable={true}
        >
          <div
            className="rc-tabs__content rc-tabs__content--animated"
          >
            <Swiper
              allowTouchMove={true}
              autoplay={false}
              autoplayInterval={300}
              className="rc-tabs__track"
              defaultIndex={0}
              indicator={false}
              loop={false}
              onIndexChange={[Function]}
              slideSize={100}
              stopPropagation={Array []}
              stuckAtBoundary={false}
              trackOffset={0}
            >
              <div
                className="rc-tabs__track rc-swiper"
                style={
                  Object {
                    "--rc-swipe-slide-size": "100%",
                    "--rc-swipe-track-offset": "0%",
                  }
                }
              >
                <div
                  className="rc-swiper__track rc-swiper__track--allow-touch-move"
                  onClickCapture={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onTouchStart={[Function]}
                >
                  <div
                    className="rc-swiper__track-inner"
                  >
                    <Animated(div)
                      className="rc-swiper__slide"
                      key=".$.$.$.$0"
                      style={
                        Object {
                          "left": "-0%",
                          "x": "0%",
                        }
                      }
                    >
                      <div
                        className="rc-swiper__slide"
                        style={
                          Object {
                            "left": "-0%",
                            "transform": "none",
                          }
                        }
                      >
                        <SwiperItem
                          className="rc-tabs__pane-wrapper"
                          key=".$.$.$0"
                          style={
                            Object {
                              "cursor": undefined,
                            }
                          }
                        >
                          <div
                            className="rc-tabs__pane-wrapper rc-swiper-item"
                            style={
                              Object {
                                "cursor": undefined,
                              }
                            }
                          >
                            <ForwardRef
                              index={0}
                              key=".$0"
                              showZeroBadge={true}
                              title="浙江省"
                              titleClass="rc-cascader__tab"
                            >
                              <div
                                className="rc-tab__pane"
                              >
                                <ul
                                  className="rc-cascader__options"
                                  key="0"
                                >
                                  <li
                                    className="rc-cascader__option rc-cascader__option--selected"
                                    key="330000"
                                    onClick={[Function]}
                                    style={
                                      Object {
                                        "color": undefined,
                                      }
                                    }
                                  >
                                    <span>
                                      浙江省
                                    </span>
                                    <Icon
                                      className="rc-cascader__selected-icon"
                                      name="success"
                                      tag="i"
                                    >
                                      <Badge
                                        className="rc-cascader__selected-icon van-icon van-icon-success"
                                        showZero={true}
                                        style={
                                          Object {
                                            "color": undefined,
                                            "fontSize": undefined,
                                          }
                                        }
                                        tag="i"
                                      >
                                        <i
                                          className="rc-badge__wrapper rc-cascader__selected-icon van-icon van-icon-success"
                                          style={
                                            Object {
                                              "color": undefined,
                                              "fontSize": undefined,
                                            }
                                          }
                                        />
                                      </Badge>
                                    </Icon>
                                  </li>
                                  <li
                                    className="rc-cascader__option"
                                    key="320000"
                                    onClick={[Function]}
                                    style={
                                      Object {
                                        "color": undefined,
                                      }
                                    }
                                  >
                                    <span>
                                      江苏省
                                    </span>
                                  </li>
                                </ul>
                              </div>
                            </ForwardRef>
                          </div>
                        </SwiperItem>
                      </div>
                    </Animated(div)>
                  </div>
                </div>
              </div>
            </Swiper>
          </div>
        </TabsContent>
      </div>
    </ForwardRef>
  </div>
</Cascader>
`;

exports[`Cascader should render title slot correctly 1`] = `"<h2 class=\\"rc-cascader__title\\">Custom Title</h2>"`;

exports[`Cascader when value change 1`] = `
<Cascader
  closeIcon="cross"
  closeable={true}
  defaultValue={
    Array [
      "330000",
    ]
  }
  options={
    Array [
      Object {
        "children": Array [
          Object {
            "text": "杭州市",
            "value": "330100",
          },
        ],
        "text": "浙江省",
        "value": "330000",
      },
      Object {
        "children": Array [
          Object {
            "text": "南京市",
            "value": "320100",
          },
        ],
        "text": "江苏省",
        "value": "320000",
      },
    ]
  }
  swipeable={true}
  value={
    Array [
      "350000",
    ]
  }
>
  <div
    className="rc-cascader"
  >
    <div
      className="rc-cascader__header"
    >
      <h2
        className="rc-cascader__title"
      />
      <Icon
        className="rc-cascader__close-icon"
        name="cross"
        onClick={[Function]}
        tag="i"
      >
        <Badge
          className="rc-cascader__close-icon van-icon van-icon-cross"
          onClick={[Function]}
          showZero={true}
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
            }
          }
          tag="i"
        >
          <i
            className="rc-badge__wrapper rc-cascader__close-icon van-icon van-icon-cross"
            onClick={[Function]}
            style={
              Object {
                "color": undefined,
                "fontSize": undefined,
              }
            }
          />
        </Badge>
      </Icon>
    </div>
    <ForwardRef
      active={0}
      animated={true}
      className="rc-cascader__tabs"
      duration={300}
      ellipsis={true}
      lazyRender={true}
      offsetTop={0}
      onClickTab={[Function]}
      swipeThreshold={0}
      swipeable={true}
      type="line"
    >
      <div
        className="rc-cascader__tabs rc-tabs rc-tabs--line"
      >
        <div
          className="rc-tabs__wrap rc-tabs__wrap--scrollable"
        >
          <div
            className="rc-tabs__nav rc-tabs__nav--line rc-tabs__nav--complete"
            role="tablist"
            style={
              Object {
                "background": undefined,
                "borderColor": undefined,
              }
            }
          >
            <ForwardRef
              className="rc-cascader__tab"
              isActive={true}
              key=".$0"
              onClick={[Function]}
              scrollable={true}
              showZeroBadge={true}
              title="浙江省"
              type="line"
            >
              <div
                aria-selected={true}
                className="rc-tab rc-tab--active rc-cascader__tab"
                onClick={[Function]}
                style={Object {}}
              >
                <span
                  className="rc-tab__text"
                >
                  浙江省
                </span>
              </div>
            </ForwardRef>
            <div
              className="rc-tabs__line"
              style={
                Object {
                  "backgroundColor": undefined,
                }
              }
            />
          </div>
        </div>
        <TabsContent
          animated={true}
          count={1}
          currentIndex={0}
          duration={300}
          inited={true}
          lazyRender={true}
          onChange={[Function]}
          swipeable={true}
        >
          <div
            className="rc-tabs__content rc-tabs__content--animated"
          >
            <Swiper
              allowTouchMove={true}
              autoplay={false}
              autoplayInterval={300}
              className="rc-tabs__track"
              defaultIndex={0}
              indicator={false}
              loop={false}
              onIndexChange={[Function]}
              slideSize={100}
              stopPropagation={Array []}
              stuckAtBoundary={false}
              trackOffset={0}
            >
              <div
                className="rc-tabs__track rc-swiper"
                style={
                  Object {
                    "--rc-swipe-slide-size": "100%",
                    "--rc-swipe-track-offset": "0%",
                  }
                }
              >
                <div
                  className="rc-swiper__track rc-swiper__track--allow-touch-move"
                  onClickCapture={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onTouchStart={[Function]}
                >
                  <div
                    className="rc-swiper__track-inner"
                  >
                    <Animated(div)
                      className="rc-swiper__slide"
                      key=".$.$.$.$0"
                      style={
                        Object {
                          "left": "-0%",
                          "x": "0%",
                        }
                      }
                    >
                      <div
                        className="rc-swiper__slide"
                        style={
                          Object {
                            "left": "-0%",
                            "transform": "none",
                          }
                        }
                      >
                        <SwiperItem
                          className="rc-tabs__pane-wrapper"
                          key=".$.$.$0"
                          style={
                            Object {
                              "cursor": undefined,
                            }
                          }
                        >
                          <div
                            className="rc-tabs__pane-wrapper rc-swiper-item"
                            style={
                              Object {
                                "cursor": undefined,
                              }
                            }
                          >
                            <ForwardRef
                              index={0}
                              key=".$0"
                              showZeroBadge={true}
                              title="浙江省"
                              titleClass="rc-cascader__tab"
                            >
                              <div
                                className="rc-tab__pane"
                              >
                                <ul
                                  className="rc-cascader__options"
                                  key="0"
                                >
                                  <li
                                    className="rc-cascader__option rc-cascader__option--selected"
                                    key="330000"
                                    onClick={[Function]}
                                    style={
                                      Object {
                                        "color": undefined,
                                      }
                                    }
                                  >
                                    <span>
                                      浙江省
                                    </span>
                                    <Icon
                                      className="rc-cascader__selected-icon"
                                      name="success"
                                      tag="i"
                                    >
                                      <Badge
                                        className="rc-cascader__selected-icon van-icon van-icon-success"
                                        showZero={true}
                                        style={
                                          Object {
                                            "color": undefined,
                                            "fontSize": undefined,
                                          }
                                        }
                                        tag="i"
                                      >
                                        <i
                                          className="rc-badge__wrapper rc-cascader__selected-icon van-icon van-icon-success"
                                          style={
                                            Object {
                                              "color": undefined,
                                              "fontSize": undefined,
                                            }
                                          }
                                        />
                                      </Badge>
                                    </Icon>
                                  </li>
                                  <li
                                    className="rc-cascader__option"
                                    key="320000"
                                    onClick={[Function]}
                                    style={
                                      Object {
                                        "color": undefined,
                                      }
                                    }
                                  >
                                    <span>
                                      江苏省
                                    </span>
                                  </li>
                                </ul>
                              </div>
                            </ForwardRef>
                          </div>
                        </SwiperItem>
                      </div>
                    </Animated(div)>
                  </div>
                </div>
              </div>
            </Swiper>
          </div>
        </TabsContent>
      </div>
    </ForwardRef>
  </div>
</Cascader>
`;
